service/resource/certificate/desired_test.go

```diff
@@ -1,9 +1,339 @@
 package certificate
 
 import (
+	"context"
+	"reflect"
 	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/spf13/afero"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/pkg/api/v1"
+
+	"github.com/giantswarm/micrologger/microloggertest"
+
+	"github.com/giantswarm/prometheus-config-controller/service/key"
+	"github.com/giantswarm/prometheus-config-controller/service/prometheus"
 )
 
 // Test_Resource_Certificate_GetDesiredState tests the GetDesiredState method.
 func Test_Resource_Certificate_GetDesiredState(t *testing.T) {
+	defaultCertificateDirectory := "/certs"
+
+	tests := []struct {
+		certificateDirectory string
+		services             []*v1.Service
+		secrets              []*v1.Secret
+
+		expectedCertificateFiles []certificateFile
+		expectedErrorHandler     func(error) bool
+	}{
+		// Test that an empty list of services leads to an empty list of certificate files.
+		{
+			certificateDirectory: defaultCertificateDirectory,
+			services:             nil,
+			secrets:              nil,
+
+			expectedCertificateFiles: []certificateFile{},
+			expectedErrorHandler:     nil,
+		},
+
+		// Test that a service without any annotation does not create a certificate file.
+		{
+			certificateDirectory: defaultCertificateDirectory,
+			services: []*v1.Service{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "foo",
+						Namespace: "default",
+					},
+				},
+			},
+			secrets: nil,
+
+			expectedCertificateFiles: []certificateFile{},
+			expectedErrorHandler:     nil,
+		},
+
+		// Test that a service with just a cluster annotation does not create a certificate file.
+		{
+			certificateDirectory: defaultCertificateDirectory,
+			services: []*v1.Service{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "apiserver",
+						Namespace: "xa5ly",
+						Annotations: map[string]string{
+							prometheus.ClusterAnnotation: "xa5ly",
+						},
+					},
+				},
+			},
+			secrets: nil,
+
+			expectedCertificateFiles: []certificateFile{},
+			expectedErrorHandler:     nil,
+		},
+
+		// Test that a service with a cluster and certificate annotation,
+		// but the certificate is missing, produces an error.
+		{
+			certificateDirectory: defaultCertificateDirectory,
+			services: []*v1.Service{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "apiserver",
+						Namespace: "xa5ly",
+						Annotations: map[string]string{
+							prometheus.ClusterAnnotation:     "xa5ly",
+							prometheus.CertificateAnnotation: "default/xa5ly-prometheus",
+						},
+					},
+				},
+			},
+			secrets: nil,
+
+			expectedCertificateFiles: nil,
+			expectedErrorHandler:     IsMissingCertificate,
```

Comment:
Not certain here. I feel like this is an error, because we've specific this certificate should exist in the service definition. There may be some cases where the secret hasn't been created _yet_, which I don't love throwing an error in, but I don't see a nicer way to do that.

tl;dr - this is probably fine, but i want another warm body to tell me it is